website/docs/server/data-loader/data-loader-instrumentation.md

```diff
@@ -0,0 +1,191 @@
+
+These custom instrumentations follow the similar approach as the default [DataLoaderDispatcherInstrumentation](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java)
+from `graphql-java`, the main difference is that regular instrumentations apply to a single `ExecutionInput` aka [GraphQL Operation](https://www.graphql-java.com/documentation/execution#queries),
+whereas these custom instrumentations apply to multiple GraphQL operations (say a BatchRequest) and stores their state in the `GraphQLContext`.
+allowing batching and deduplication of transactions across those multiple GraphQL operations.
+
+By default, each GraphQL operation is processed independently of each other. Multiple operations can be processed
+together as if they were single GraphQL request if they are part of the same batch request.
+
+The `graphql-kotlin-dataloader-instrumentation` module contains 2 custom `DataLoader` instrumentations:
+
+1. **DataLoaderLevelDispatchedInstrumentation**
+
+   Dispatches all data loaders when a certain **Level** of all **ExecutionInputs** sharing a
+   **GraphQLContext** was dispatched (all DataFetchers were invoked).
+
+2. **DataLoaderSyncExecutionExhaustedInstrumentation**
+
+   Dispatches all data loaders when all synchronous execution paths across all operations are exhausted. Synchronous execution path
+   is considered exhausted when all currently processed data fetchers were either resolved to a scalar or dispatched a `CompletableFuture`.
+
+## Dispatching by level
+
+The `DataLoaderLevelDispatchedInstrumentation` tracks the state of all `ExecutionInputs` across operations. When a certain
+field dispatches, it will check if all fields across all operations for a particular level were dispatched and if the condition is met,
+it will dispatch all the data loaders.
+
+### Example
+
+**Queries**
+
+```graphql
+query Q1 {
+    astronaut(id: 1) { # async
+        id
+        name
+        missions { # async
+            id
+            designation
+        }
+    }
+}
+
+query Q2 {
+    astronaut(id: 2) { # async
+        id
+        name
+        missions { # async
+            id
+            designation
+        }
+    }
+}
+```
+
+**Execution diagram**
+
+![Image of data loader level dispatched instrumentation](../../assets/data-loader-level-dispatched-instrumentation.png)
+
+* The `astronaut` `DataFetcher` uses a `AstronautDataLoader` which will be dispatched when **Level 1** of those 2 operations
+  is dispatched, causing the `AstronautDataLoader` to load 2 astronauts.
+* The `missions` `DataFetcher` uses a `MissionsByAstronautDataLoader` which will be dispatched when **Level 2** of those 2 operations
+  is dispatched, causing the `MissionsByAstronautDataLoader` to load 2 lists of missions by astronaut.
+
+### Usage on spring server
```

Comment:
I think we should provide docs on integration outside of spring server (there could be a note below saying that our spring support simplifies the above setup and you can use this simple yaml configuration to enable it). Guess you don't need to specify the complete schema setup part -> main part is to highlight that this instrumentation requires population of GraphQL context map with the state (this comment applies to both instrumentations).

i.e. something along the lines

===
In order to enable batching by level, you need to configure your GraphQL instance with the `DataLoaderLevelDispatchedInstrumentation`

```kotlin
val graphQL = GraphQL
    .instrumentation(DataLoaderLevelDispatchedInstrumentation())
    // configure schema, type wiring, etc.
    .build()
```

This data loader instrumentation relies on a global state object that should be stored in the GraphQL context map

```kotlin
val graphQLContext = mapOf(
    KotlinDataLoaderRegistry::class to kotlinDataLoaderRegistry,
    SyncExecutionExhaustedState::class to ExecutionLevelDispatchedState(queries.size)
)
```

:::info
`graphql-kotlin-spring-server` provides convenient integration of batch loader functionality through simple configuration. Batching instrumentation can be enabled by configuring following properties

```yaml
graphql:
  batching:
   enabled: true
   strategy: LEVEL_DISPATCHED
```
:::